File: docs/absorbed-from-phenotype-gfx/unity/terrain/src/HeightField.cs
using System;

namespace Phenotype.Terrain
{
    /// <summary>
    /// Stores and queries per-tile elevation data for terrain height sampling.
    /// Consumers call into this to resolve world-space Y for a given tile coordinate.
    /// </summary>
    /// <remarks>
    /// The height field stores data in a flat float array indexed as <c>z * Width + x</c>.
    /// Coordinates are zero-based and range from <c>(0, 0)</c> to <c>(Width - 1, Height - 1)</c>.
    /// </remarks>
    /// <example>
    /// <code>
    /// var hf = new HeightField(256, 256);
    /// hf.SetHeight(10, 10, 42.0f);
    /// float y = hf.GetHeight(10, 10); // 42.0f
    /// </code>
    /// </example>
    public class HeightField
    {
        private readonly float[] _data;

        /// <summary>
        /// Width of the height field in tiles.
        /// </summary>
        public int Width { get; }

        /// <summary>
        /// Height of the height field in tiles.
        /// </summary>
        public int Height { get; }

        /// <summary>
        /// Creates a new height field with the specified dimensions and optional elevation data.
        /// </summary>
        /// <param name="width">Number of tiles along the X axis. Must be non-negative.</param>
        /// <param name="height">Number of tiles along the Z axis. Must be non-negative.</param>
        /// <param name="data">Optional flat array of elevation values. When null, a zero-initialized array is used.</param>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when width or height is negative.</exception>
        /// <exception cref="ArgumentException">Thrown when data length does not match width * height.</exception>
        /// <example>
        /// <code>
        /// // Create a flat 64x64 height field
        /// var hf = new HeightField(64, 64);
        ///
        /// // Create from existing data
        /// float[] data = new float[256];
        /// var hf2 = new HeightField(16, 16, data);
        /// </code>
        /// </example>
        public HeightField(int width, int height, float[] data = null)
        {
            if (width < 0)
                throw new ArgumentOutOfRangeException(nameof(width), "Width must be non-negative.");
            if (height < 0)
                throw new ArgumentOutOfRangeException(nameof(height), "Height must be non-negative.");

            int expectedLength = width * height;
            if (data != null && data.Length != expectedLength)
                throw new ArgumentException(
                    $"Data length ({data.Length}) does not match expected size ({expectedLength}).",
                    nameof(data));

            Width = width;
            Height = height;
            _data = data ?? new float[expectedLength];
        }

        /// <summary>
        /// Returns the elevation at the given tile coordinate.
        /// </summary>
        /// <param name="x">Tile coordinate along the X axis. Must be within <c>[0, Width)</c>.</param>
        /// <param name="z">Tile coordinate along the Z axis. Must be within <c>[0, Height)</c>.</param>
        /// <returns>Elevation value in world units.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when coordinates are out of bounds.</exception>
        /// <example>
        /// <code>
        /// var hf = new HeightField(64, 64);
        /// hf.SetHeight(5, 5, 10.0f);
        /// float elevation = hf.GetHeight(5, 5); // 10.0f
        /// </code>
        /// </example>
        public float GetHeight(int x, int z)
        {
            if (x < 0 || x >= Width)
                throw new ArgumentOutOfRangeException(nameof(x), "X coordinate is out of bounds.");
            if (z < 0 || z >= Height)
                throw new ArgumentOutOfRangeException(nameof(z), "Z coordinate is out of bounds.");

            return _data[z * Width + x];
        }

        /// <summary>
        /// Sets the elevation at the given tile coordinate.
        /// </summary>
        /// <param name="x">Tile coordinate along the X axis. Must be within <c>[0, Width)</c>.</param>
        /// <param name="z">Tile coordinate along the Z axis. Must be within <c>[0, Height)</c>.</param>
        /// <param name="value">Elevation value in world units.</param>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when coordinates are out of bounds.</exception>
        /// <example>
        /// <code>
        /// var hf = new HeightField(64, 64);
        /// hf.SetHeight(0, 0, 100.0f);
        /// hf.SetHeight(63, 63, 50.0f);
        /// </code>
        /// </example>
        public void SetHeight(int x, int z, float value)
        {
            if (x < 0 || x >= Width)
                throw new ArgumentOutOfRangeException(nameof(x), "X coordinate is out of bounds.");
            if (z < 0 || z >= Height)
                throw new ArgumentOutOfRangeException(nameof(z), "Z coordinate is out of bounds.");

            _data[z * Width + x] = value;
        }

        /// <summary>
        /// Returns a copy of the internal elevation data array.
        /// </summary>
        /// <returns>A new array containing the current elevation values in row-major order (z * Width + x).</returns>
        /// <example>
        /// <code>
        /// var hf = new HeightField(4, 4);
        /// hf.SetHeight(1, 2, 5.0f);
        /// float[] snapshot = hf.GetData();
        /// // snapshot[2 * 4 + 1] == 5.0f
        /// </code>
        /// </example>
        public float[] GetData()
        {
            float[] copy = new float[_data.Length];
            _data.CopyTo(copy, 0);
            return copy;
        }
    }
}

File: docs/absorbed-from-phenotype-gfx/src/terrain/height_field.rs
//! 2D elevation grid.
//!
//! Ported from C# `HeightField.cs` (137 lines). Stores a flat `Vec<f32>`
//! indexed as `z * width + x`. World coordinates are 32-bit signed tile
//! indices; elevations are `f32` world units.

use serde::{Deserialize, Serialize};

use crate::terrain::error::{TerrainError, TerrainResult};

/// 2D elevation field indexed as `z * width + x`.
///
/// Coordinates are zero-based and range from `(0, 0)` to `(width - 1, height - 1)`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HeightField {
    width: i32,
    height: i32,
    data: Vec<f32>,
}

impl HeightField {
    /// Create a new height field with the given dimensions and zero-initialized data.
    pub fn new(width: i32, height: i32) -> TerrainResult<Self> {
        Self::with_data(width, height, None)
    }

    /// Create with optional initial data. `data.len()` must equal `width * height`
    /// when supplied.
    pub fn with_data(width: i32, height: i32, data: Option<Vec<f32>>) -> TerrainResult<Self> {
        if width < 0 {
            return Err(TerrainError::OutOfBounds {
                msg: format!("width must be non-negative, got {width}"),
            });
        }
        if height < 0 {
            return Err(TerrainError::OutOfBounds {
                msg: format!("height must be non-negative, got {height}"),
            });
        }
        let expected = (width as usize).saturating_mul(height as usize);
        let data = match data {
            Some(d) => {
                if d.len() != expected {
                    return Err(TerrainError::InvalidDataLength {
                        got: d.len(),
                        expected,
                    });
                }
                d
            }
            None => vec![0.0; expected],
        };
        Ok(Self { width, height, data })
    }

    /// Width of the height field in tiles.
    pub fn width(&self) -> i32 { self.width }

    /// Height of the height field in tiles.
    pub fn height(&self) -> i32 { self.height }

    /// Returns the elevation at the given tile coordinate.
    pub fn get_height(&self, x: i32, z: i32) -> TerrainResult<f32> {
        if x < 0 || x >= self.width {
            return Err(TerrainError::OutOfBounds { msg: format!("x={x} out of [0, {})", self.width) });
        }
        if z < 0 || z >= self.height {
            return Err(TerrainError::OutOfBounds { msg: format!("z={z} out of [0, {})", self.height) });
        }
        Ok(self.data[z as usize * self.width as usize + x as usize])
    }

    /// Sets the elevation at the given tile coordinate.
    pub fn set_height(&mut self, x: i32, z: i32, value: f32) -> TerrainResult<()> {
        if x < 0 || x >= self.width {
            return Err(TerrainError::OutOfBounds { msg: format!("x={x} out of [0, {})", self.width) });
        }
        if z < 0 || z >= self.height {
            return Err(TerrainError::OutOfBounds { msg: format!("z={z} out of [0, {})", self.height) });
        }
        self.data[z as usize * self.width as usize + x as usize] = value;
        Ok(())
    }

    /// Returns a copy of the internal elevation data.
    pub fn get_data(&self) -> Vec<f32> {
        self.data.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn heightfield_can_be_instantiated() {
        let hf = HeightField::new(16, 16).unwrap();
        assert_eq!(hf.width(), 16);
        assert_eq!(hf.height(), 16);
    }

    #[test]
    fn heightfield_get_height_returns_zero_for_default_data() {
        let hf = HeightField::new(4, 4).unwrap();
        assert_eq!(hf.get_height(0, 0).unwrap(), 0.0);
        assert_eq!(hf.get_height(3, 3).unwrap(), 0.0);
    }

    #[test]
    fn heightfield_set_get_round_trip() {
        let mut hf = HeightField::new(4, 4).unwrap();
        hf.set_height(1, 2, 42.5).unwrap();
        assert_eq!(hf.get_height(1, 2).unwrap(), 42.5);
    }

    #[test]
    fn heightfield_get_data_returns_copy() {
        let mut hf = HeightField::new(2, 2).unwrap();
        hf.set_height(0, 0, 5.0).unwrap();
        let mut data = hf.get_data();
        data[0] = 99.0;
        // original unchanged because get_data cloned
        assert_eq!(hf.get_height(0, 0).unwrap(), 5.0);
    }

    // ---- Edge cases (from HeightFieldEdgeCaseTests) ----

    #[test]
    fn heightfield_negative_height_round_trip() {
        let mut hf = HeightField::new(4, 4).unwrap();
        hf.set_height(1, 1, -42.5).unwrap();
        hf.set_height(2, 3, -999.99).unwrap();
        assert_eq!(hf.get_height(1, 1).unwrap(), -42.5);
        assert_eq!(hf.get_height(2, 3).unwrap(), -999.99);
    }

    #[test]
    fn heightfield_negative_heights_via_data_array() {
        let data = vec![0.0, -1.0, -2.0, -3.0];
        let hf = HeightField::with_data(2, 2, Some(data)).unwrap();
        assert_eq!(hf.get_height(1, 0).unwrap(), -1.0);
        assert_eq!(hf.get_height(0, 1).unwrap(), -2.0);
        assert_eq!(hf.get_height(1, 1).unwrap(), -3.0);
    }

    #[test]
    fn heightfield_zero_size_can_be_instantiated() {
        let hf = HeightField::new(0, 0).unwrap();
        assert_eq!(hf.width(), 0);
        assert_eq!(hf.height(), 0);
    }

    #[test]
    fn heightfield_zero_size_get_height_throws() {
        let hf = HeightField::new(0, 0).unwrap();
        assert!(hf.get_height(0, 0).is_err());
    }

    #[test]
    fn heightfield_zero_size_set_height_throws() {
        let mut hf = HeightField::new(0, 0).unwrap();
        assert!(hf.set_height(0, 0, 1.0).is_err());
    }

    #[test]
    fn heightfield_large_1024_can_be_instantiated() {
        let hf = HeightField::new(1024, 1024).unwrap();
        assert_eq!(hf.width(), 1024);
        assert_eq!(hf.height(), 1024);
    }

    #[test]
    fn heightfield_large_1024_read_write_round_trip() {
        let mut hf = HeightField::new(1024, 1024).unwrap();
        hf.set_height(0, 0, 1.0).unwrap();
        hf.set_height(1023, 1023, 2.0).unwrap();
        hf.set_height(512, 512, 3.0).unwrap();
        assert_eq!(hf.get_height(0, 0).unwrap(), 1.0);
        assert_eq!(hf.get_height(1023, 1023).unwrap(), 2.0);
        assert_eq!(hf.get_height(512, 512).unwrap(), 3.0);
    }

    #[test]
    fn heightfield_invalid_data_length() {
        let result = HeightField::with_data(2, 2, Some(vec![0.0, 1.0]));
        assert!(matches!(result, Err(TerrainError::InvalidDataLength { .. })));
    }
}
